README.md
# NAME

GitInsight - Predict your github contributions using Bayesian inference and Markov chain

# SYNOPSIS

    gitinsight --username [githubusername] (--nodaystats) (--accuracy) #using the shipped bin

    #or using the module

    my $Insight= GitInsight->new(no_day_stats => 0, username => "markov", accuracy=> 1);
    my $Result= $Insight->process;
    my $accuracy = $Insight->{accuracy};
    $Result = $Insight->{result};
    # $Result contains the next week predictions and is an arrayref of arrayrefs    [  [ 'Sat', 1, '2014-07-1', [ 0 ,  '0.151515151515152', '0.0606060606060606', '0.0404040404040404',  0  ]  ],   ..   [            'DayofWeek',      'winner_label',  'day' ,  [             probability_label_0,  probability_label_1,              probability_label_2,          probability_label_3,              probability_label_4            ]          ]]

# DESCRIPTION

GitInsight is module that allow you to predict your github contributions in the "calendar contribution" style of github (the table of contribution that you see on your profile page).

# HOW DOES IT WORK?

GitInsight generates a transation probrability matrix from your github contrib\_calendar to compute the possibles states for the following days. Given that GitHub split the states thru 5 states (or here also called label), the probability can be inferenced by using Bayesian methods to update the beliefs of the possible state transition, while markov chain is used to predict the states. The output of the submitted data is then plotted using Cellular Automata.

## THEORY

We trace the transitions states in a matrix and increasing the count as far as we observe a transition ([https://en.wikipedia.org/wiki/Transition\_matrix](https://en.wikipedia.org/wiki/Transition_matrix)), then we inference the probabilities using Bayesan method [https://en.wikipedia.org/wiki/Bayesian\_inference](https://en.wikipedia.org/wiki/Bayesian_inference) [https://en.wikipedia.org/wiki/Examples\_of\_Markov\_chains](https://en.wikipedia.org/wiki/Examples_of_Markov_chains).

# INSTALLATION

GitInsight requires the installation of gsl (GNU scientific library), gd(http://libgd.org/), PDL and PDL::Stats  (to be installed after the gsl library set).

on Debian:

        apt-get install gsl-bin libgs10-devt
        apt-get install pdl libpdl-stats-perl libgd2-xpm-dev

It's reccomended to use cpanm to install all the required deps, install it thru your package manager or just do:

    cpan App::cpanminus

After the installation of gsl, clone the repository and install all the dependencies with cpanm:

    cpanm --installdeps .

Then install it as usual:

    perl Build.PL
    ./Build
    ./Build test #ensure that the module works correctly
    ./Build install

# OPTIONS

## username

required, it's the GitHub username used to calculate the prediction

## ca\_output

you can enable/disable the cellular autmata output using this option (1/0)

## no\_day\_stats

setting this option to 1, will slightly change the prediction: it will be calculated a unique transition matrix instead one for each day

## left\_cutoff

used to cut the days from the start (e.g. if you want to delete the first 20 days from the prediction, just set this to 20)

## cutoff\_offset

used to select a range where the prediction happens (e.g. if you want to calculate the prediction of a portion of your year of contribution)

## file\_output

here you can choose the file output name for ca.

## accuracy

Enable/disable accuracy calculation (1/0)

## verbose

Enable/disable verbosity (1/0)

# METHODS

## contrib\_calendar($username)

Fetches the github contrib\_calendar of the specified user

## process

Calculate the predictions and generate the CA

## start\_day

Returns the first day of the contrib\_calendar

## last\_day

Returns the last day of the contrib calendar (prediction included)

## prediction\_start\_day

Returns the first day of the prediction (7 days of predictions)

# AUTHOR

mudler <mudler@dark-lab.net>

# COPYRIGHT

Copyright 2014- mudler

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

[GitInsight::Util](https://metacpan.org/pod/GitInsight::Util), [PDL](https://metacpan.org/pod/PDL), [PDL::Stats](https://metacpan.org/pod/PDL::Stats)


Changes
Revision history for GitInsight

{{$NEXT}}

0.06  2014-11-03 20:00:57 CET

	- Fixing POD

0.05  2014-11-03 19:51:41 CET

	- updated META, added git repo

0.04  2014-11-03 18:27:35 CET
	- Added accuracy computation


lib/GitInsight/Obj.pm
package GitInsight::Obj;

 use strict;
 use warnings;
 use utf8;

 #use feature ();

 our $feature = eval {
     require feature;
     feature->import();
     1;
 };


 # Only Perl 5.14+ requires it on demand
 use IO::Handle ();

 # Protect subclasses using AUTOLOAD
 sub DESTROY { }

 sub import {
     my $class = shift;
     return unless my $flag = shift;

     # Base
     if ( $flag eq '-base' ) { $flag = $class }

     # Strict
     elsif ( $flag eq '-strict' ) { $flag = undef }

     # Module
     elsif ( ( my $file = $flag ) && !$flag->can('new') ) {
         $file =~ s!::|'!/!g;
         require "$file.pm";
     }

     # ISA
     if ($flag) {
         my $caller = caller;
         no strict 'refs';
         push @{"${caller}::ISA"}, $flag;
         *{"${caller}::has"} = sub { attr( $caller, @_ ) };
     }

     # Mojo modules are strict!
     $_->import for qw(strict warnings utf8);
     if ($feature) {
         feature->import(':5.10');
     }
 }

 sub attr {
     my ( $self, $attrs, $default ) = @_;
     return unless ( my $class = ref $self || $self ) && $attrs;

     die 'Default has to be a code reference or constant value'
         if ref $default && ref $default ne 'CODE';

     for my $attr ( @{ ref $attrs eq 'ARRAY' ? $attrs : [$attrs] } ) {
         die qq{Attribute "$attr" invalid}
             unless $attr =~ /^[a-zA-Z_]\w*$/;

         # Header (check arguments)
         my $code = "package $class;\nsub $attr {\n  if (\@_ == 1) {\n";

         # No default value (return value)
         unless ( defined $default ) { $code .= "    return \$_[0]{'$attr'};" }

         # Default value
         else {

             # Return value
             $code
                 .= "    return \$_[0]{'$attr'} if exists \$_[0]{'$attr'};\n";

             # Return default value
             $code .= "    return \$_[0]{'$attr'} = ";
             $code .=
                 ref $default eq 'CODE'
                 ? '$default->($_[0]);'
                 : '$default;';
         }

         # Store value
         $code .= "\n  }\n  \$_[0]{'$attr'} = \$_[1];\n";

         # Footer (return invocant)
         $code .= "  \$_[0];\n}";

         warn "-- Attribute $attr in $class\n$code\n\n"
             if $ENV{GitInsight_OBJ_DEBUG};
         die "GitInsight::Obj error: $@" unless eval "$code;1";
     }
 }

 sub new {
     my $class = shift;
     bless @_ ? @_ > 1 ? {@_} : { %{ $_[0] } } : {}, ref $class || $class;
 }

 sub tap {
     my ( $self, $cb ) = @_;
     $_->$cb for $self;
     return $self;
 }

 1;
 __END__

 =encoding utf8

 =head1 NAME

 GitInsight::Obj - Minimal base class for GitInsight

 =head1 SYNOPSIS

   package Cat;
   use GitInsight::Obj -base;

   has name => 'Nyan';
   has [qw(birds mice)] => 2;

   package Tiger;
   use GitInsight::Obj 'Cat';

   has friend  => sub { Cat->new };
   has stripes => 42;

   package main;
   use GitInsight::Obj -strict;

   my $mew = Cat->new(name => 'Longcat');
   say $mew->mice;
   say $mew->mice(3)->birds(4)->mice;

   my $rawr = Tiger->new(stripes => 23, mice => 0);
   say $rawr->tap(sub { $_->friend->name('Tacgnol') })->mice;

 =head1 DESCRIPTION

 L<GitInsight::Obj> is a simple base class for L<GitInsight>, a fork of  L<Mojo::Base>.

   # Automatically enables "strict", "warnings", "utf8" and Perl 5.10 features
   use GitInsight::Obj -strict;
   use GitInsight::Obj -base;
   use GitInsight::Obj 'SomeBaseClass';

 All three forms save a lot of typing.

   # use GitInsight::Obj -strict;
   use strict;
   use warnings;
   use utf8;
   use feature ':5.10';
   use IO::Handle ();

   # use GitInsight::Obj -base;
   use strict;
   use warnings;
   use utf8;
   use feature ':5.10';
   use IO::Handle ();
   use GitInsight::Obj;
   push @ISA, 'GitInsight::Obj';
   sub has { GitInsight::Obj::attr(__PACKAGE__, @_) }

   # use GitInsight::Obj 'SomeBaseClass';
   use strict;
   use warnings;
   use utf8;
   use feature ':5.10';
   use IO::Handle ();
   require SomeBaseClass;
   push @ISA, 'SomeBaseClass';
   use GitInsight::Obj;
   sub has { GitInsight::Obj::attr(__PACKAGE__, @_) }

 =head1 FUNCTIONS

 L<GitInsight::Obj> implements the following functions like L<Mojo::Base>, which can be imported with
 the C<-base> flag or by setting a base class.

 =head2 has

   has 'name';
   has [qw(name1 name2 name3)];
   has name => 'foo';
   has name => sub {...};
   has [qw(name1 name2 name3)] => 'foo';
   has [qw(name1 name2 name3)] => sub {...};

 Create attributes for hash-based objects, just like the L</"attr"> method.

 =head1 METHODS

 L<GitInsight::Obj> implements the following methods.

 =head2 attr

   $object->attr('name');
   BaseSubClass->attr('name');
   BaseSubClass->attr([qw(name1 name2 name3)]);
   BaseSubClass->attr(name => 'foo');
   BaseSubClass->attr(name => sub {...});
   BaseSubClass->attr([qw(name1 name2 name3)] => 'foo');
   BaseSubClass->attr([qw(name1 name2 name3)] => sub {...});

 Create attribute accessor for hash-based objects, an array reference can be
 used to create more than one at a time. Pass an optional second argument to
 set a default value, it should be a constant or a callback. The callback will
 be executed at accessor read time if there's no set value. Accessors can be
 chained, that means they return their invocant when they are called with an
 argument.

 =head2 new

   my $object = BaseSubClass->new;
   my $object = BaseSubClass->new(name => 'value');
   my $object = BaseSubClass->new({name => 'value'});

 This base class provides a basic constructor for hash-based objects. You can
 pass it either a hash or a hash reference with attribute values.

 =head2 tap

   $object = $object->tap(sub {...});

 K combinator, tap into a method chain to perform operations on an object
 within the chain. The object will be the first argument passed to the callback
 and is also available as C<$_>.

 =head1 DEBUGGING

 You can set the C<GitInsight_OBJ_DEBUG> environment variable to get some advanced
 diagnostics information printed to C<STDERR>.

   GitInsight_OBJ_DEBUG=1

 =head1 SEE ALSO

 L<GitInsight>, L<Mojo::Base>.

 =cut


lib/GitInsight/Loader.pm
package GitInsight::Loader;
 use GitInsight::Base -base;
 
 use File::Basename 'fileparse';
 use File::Spec::Functions qw(catdir catfile splitdir);
 use GitInsight::Exception;
 use GitInsight::Util qw(b64_decode class_to_path);
 
 my (%BIN, %CACHE);
 
 sub data { $_[1] ? $_[2] ? _all($_[1])->{$_[2]} : _all($_[1]) : undef }
 
 sub is_binary { keys %{_all($_[1])} ? !!$BIN{$_[1]}{$_[2]} : undef }
 
 sub load {
   my ($self, $module) = @_;
 
   # Check module name
   return 1 if !$module || $module !~ /^\w(?:[\w:']*\w)?$/;
 
   # Load
   return undef if $module->can('new') || eval "require $module; 1";
 
   # Exists
   return 1 if $@ =~ /^Can't locate \Q@{[class_to_path $module]}\E in \@INC/;
 
   # Real error
   return GitInsight::Exception->new($@);
 }
 
 sub search {
   my ($self, $ns) = @_;
 
   my %modules;
   for my $directory (@INC) {
     next unless -d (my $path = catdir $directory, split(/::|'/, $ns));
 
     # List "*.pm" files in directory
     opendir(my $dir, $path);
     for my $file (grep /\.pm$/, readdir $dir) {
       next if -d catfile splitdir($path), $file;
       $modules{"${ns}::" . fileparse $file, qr/\.pm/}++;
     }
   }
 
   return [keys %modules];
 }
 
 sub _all {
   my $class = shift;
 
   return $CACHE{$class} if $CACHE{$class};
   my $handle = do { no strict 'refs'; \*{"${class}::DATA"} };
   return {} unless fileno $handle;
   seek $handle, 0, 0;
   my $data = join '', <$handle>;
 
   # Ignore everything before __DATA__ (Windows will seek to start of file)
   $data =~ s/^.*\n__DATA__\r?\n/\n/s;
 
   # Ignore everything after __END__
   $data =~ s/\n__END__\r?\n.*$/\n/s;
 
   # Split files
   (undef, my @files) = split /^@@\s*(.+?)\s*\r?\n/m, $data;
 
   # Find data
   my $all = $CACHE{$class} = {};
   while (@files) {
     my ($name, $data) = splice @files, 0, 2;
     $all->{$name} = $name =~ s/\s*\(\s*base64\s*\)$//
       && ++$BIN{$class}{$name} ? b64_decode($data) : $data;
   }
 
   return $all;
 }
 
 1;
 
 =encoding utf8
 
 =head1 NAME
 
 GitInsight::Loader - Loader
 
 =head1 SYNOPSIS
 
   use GitInsight::Loader;
 
   # Find modules in a namespace
   my $loader = GitInsight::Loader->new;
   for my $module (@{$loader->search('Some::Namespace')}) {
 
     # Load them safely
     my $e = $loader->load($module);
     warn qq{Loading "$module" failed: $e} and next if ref $e;
 
     # And extract files from the DATA section
     say $loader->data($module, 'some_file.txt');
   }
 
 =head1 DESCRIPTION
 
 L<GitInsight::Loader> is a class loader and plugin framework.
 
 =head1 METHODS
 
 L<GitInsight::Loader> inherits all methods from L<GitInsight::Base> and implements the
 following new ones.
 
 =head2 data
 
   my $all   = $loader->data('Foo::Bar');
   my $index = $loader->data('Foo::Bar', 'index.html');
 
 Extract embedded file from the C<DATA> section of a class, all files will be
 cached once they have been accessed for the first time.
 
   say for keys %{$loader->data('Foo::Bar')};
 
 =head2 is_binary
 
   my $bool = $loader->is_binary('Foo::Bar', 'test.png');
 
 Check if embedded file from the C<DATA> section of a class was Base64 encoded.
 
 =head2 load
 
   my $e = $loader->load('Foo::Bar');
 
 Load a class and catch exceptions. Note that classes are checked for a C<new>
 method to see if they are already loaded.
 
   if (my $e = $loader->load('Foo::Bar')) {
     die ref $e ? "Exception: $e" : 'Not found!';
   }
 
 =head2 search
 
   my $modules = $loader->search('MyApp::Namespace');
 
 Search for modules in a namespace non-recursively.
 
 =head1 SEE ALSO
 
 L<GitInsight::Obj>, L<Mojo::Base>.
 
 =cut

lib/GitInsight.pm
package GitInsight;

# XXX: Add behavioural change detection, focusing on that period for predictions

BEGIN {
    $|  = 1;
    $^W = 1;
}
our $VERSION = '0.05';

#use Carp::Always;
use GitInsight::Obj -base;
use strict;
use warnings;
use 5.008_005;
use GD::Simple;

use Carp;
use Storable qw(dclone);
use POSIX;
use Time::Local;
use GitInsight::Util
    qw(markov markov_list LABEL_DIM gen_m_mat gen_trans_mat info error warning wday label prob label_step);
use List::Util qw(max);

use LWP::UserAgent;
use POSIX qw(strftime ceil);

has [qw(username contribs calendar)];
has 'verbose'      => sub {0};
has 'no_day_stats' => sub {0};
has 'statistics'   => sub {0};
has 'ca_output'    => sub {1};
has 'accuracy'     => sub {0};
has [qw(left_cutoff cutoff_offset file_output)];

sub contrib_calendar {
    my $self = shift;
    my $username = shift || $self->username;
    $self->username($username) if !$self->username;
    my $ua = LWP::UserAgent->new;
    $ua->timeout(10);
    $ua->env_proxy;
    my $response
        = $ua->get(
        'https://github.com/users/' . $username . '/contributions' );
    info "Getting "
        . 'https://github.com/users/'
        . $username
        . '/contributions'
        if $self->verbose;

    if ( $response->is_success ) {
        $self->decode( $response->decoded_content );
        return $self->contribs;
    }
    else {
        die $response->status_line;
    }
}

sub draw_ca {
    my $self = shift;
    my @CA   = @_;
    my $cols = ceil( $#CA / 7 ) + 1;
    my $rows = 7;

    my $cell_width  = 50;
    my $cell_height = 50;
    my $border      = 3;
    my $width       = $cols * $cell_width;
    my $height      = $rows * $cell_height;

    my $img = GD::Simple->new( $width, $height );

    $img->font(gdSmallFont);    #i'll need that later
    for ( my $c = 0; $c < $cols; $c++ ) {
        for ( my $r = 0; $r < $rows; $r++ ) {
            my $color = $CA[ $c * $rows + $r ]
                or
                next; #infering ca from sequences of colours generated earlier
            my @topleft = ( $c * $cell_width, $r * $cell_height );
            my @botright = (
                $topleft[0] + $cell_width - $border,
                $topleft[1] + $cell_height - $border
            );
            eval {
                $img->bgcolor( @{$color} );
                $img->fgcolor( @{$color} );
            };
            $img->rectangle( @topleft, @botright );
            $img->moveTo( $topleft[0] + 2, $botright[1] + 2 );
            $img->fgcolor( 255, 0, 0 )
                and $img->rectangle( @topleft, @botright )
                if ( $c * $rows + $r >= ( scalar(@CA) - 7 ) );
            $img->fgcolor( 0, 0, 0 )
                and $img->string( $GitInsight::Util::wday[$r] )
                if ( $c == 0 );
        }
    }
    if ( defined $self->file_output ) {
        my $filename = $self->file_output . ".png";

        #. "/"
        #. join( "_", $self->start_day, $self->last_day ) . "_"
        #. $self->username . "_"
        #. scalar(@CA) .
        open my $PNG, ">" . $filename;
        binmode($PNG);
        print $PNG $img->png;
        close $PNG;
        info "File written in : " . $filename if $self->verbose;
        return $filename;
    }
    else {
        return $img->png;
    }

}

# useful when interrogating the object
sub start_day            { shift->{first_day}->{data} }
sub last_day             { @{ shift->{result} }[-1]->[2] }
sub prediction_start_day { @{ shift->{result} }[0]->[2] }

sub _accuracy {
    my $self = shift;
    my ( @chunks, @commits );
    push @chunks, [ splice @{ $self->calendar }, 0, 7 ]
        while @{ $self->calendar };

    #@chunks contain a list of arrays of 7 days each

    my $total_days = 0;
    my $accuracy   = 0;
    for (@chunks) {

        # next if @{$_} < 4;
        push( @commits, @{$_} );
        my $Insight = GitInsight->new(
            no_day_stats => $self->no_day_stats,
            ca_output    => 0,
            username     => $self->username
        );    #disable png generation
        $Insight->decode( [@commits] )
            ;    #using $_ for small contributors is better
        $Insight->process;
        foreach my $res ( @{ $Insight->{result} } ) {
            next if ( !exists $self->contribs->{ $res->[2] }->{l} );
            $accuracy++
                if ( $self->contribs->{ $res->[2] }->{l} == $res->[1] );
            $total_days++;
        }
    }
    my $accuracy_prob = prob( $total_days, $accuracy );
    $self->{accuracy} = $accuracy_prob;
    info "Accuracy is $accuracy / $total_days" if $self->verbose;
    info sprintf( "%.5f", $accuracy_prob * 100 ) . " \%" if $self->verbose;
    return $self;
}

sub _decode_calendar {
    shift;
    my $content = shift;
    my @out;
    push( @out, [ $2, $1 ] )
        while ( $content =~ m/data\-count="(.*?)" data\-date="(.*?)"/g );

    return \@out;
}

# first argument is the data:
# it should be a string in the form [ [2013-01-20, 9], ....    ] a stringified form of arrayref. each element must be an array ref containing in the first position the date, and in the second the commits .
sub decode {
    my $self = shift;

    #my $response = ref $_[0] ne "ARRAY" ? eval(shift) : shift;
    my $response
        = ref $_[0] ne "ARRAY" ? $self->_decode_calendar(shift) : shift;
    $self->calendar( dclone($response) );
    my %commits_count;
    my $min = $self->left_cutoff || 0;
    $self->{result} = [];    #empty the result
    $min = 0 if ( $min < 0 );    # avoid negative numbers
    my $max
        = $self->cutoff_offset || ( scalar( @{$response} ) - 1 );
    $max = scalar( @{$response} )
        if $max > scalar( @{$response} )
        ;    # maximum cutoff boundary it's array element number
    info "$min -> $max portion" if $self->verbose;
    my $max_commit
        = max( map { $_->[1] } @{$response} );    #Calculating label steps
    label_step( 0 .. $max_commit );   #calculating quartiles over commit count
    info( "Max commit is: " . $max_commit ) if $self->verbose;
    $self->{first_day}->{day} = wday( $response->[0]->[0] )
        ; #getting the first day of the commit calendar, it's where the ca will start
    my ($index)
        = grep { $GitInsight::Util::wday[$_] eq $self->{first_day}->{day} }
        0 .. $#GitInsight::Util::wday;
    $self->{first_day}->{index} = $index;
    $self->{first_day}->{data}  = $response->[$min]->[0];
    push( @{ $self->{ca} }, [ 255, 255, 255 ] )
        for (
        0 .. scalar(@GitInsight::Util::wday)    #white fill for labels
        + ( $index - 1 )
        );                                      #white fill for no contribs

    $self->{transition} = gen_trans_mat( $self->no_day_stats );
    my $last;
    $self->{last_week}
        = [ map { [ $_->[0], label( $_->[1] ) ] }
            ( @{$response} )[ ( ( $max + $min ) - 6 ) .. ( $max + $min ) ] ]
        ; # cutting the last week from the answer and substituting the label instead of the commit number
          #print( $self->{transition}->{$_} ) for (last_week keys $self->{transition} );
          # $self->{max_commit} =0;
    $self->contribs(
        $self->no_day_stats
        ? { map {
                my $l = label( $_->[1] );
                push( @{ $self->{ca} }, $GitInsight::Util::CA_COLOURS{$l} )
                    ;    #building the ca
                $last = $l if ( !$last );

                #    $commits_count{ $_->[1] } = 1;
                $self->{stats}->{$l}++
                    if $self->statistics == 1;    #filling stats hashref
                $self->{transition_hash}->{$last}->{$l}++
                    ; #filling transition_hash hashref from $last (last seen label) to current label
                $self->{transition}
                    ->slice("$last,$l")++;    #filling transition matrix
                 #$self->{max_commit} = $_->[1] if ($_->[1]>$self->{max_commit});
                $last = $l;
                $_->[0] => {
                    c => $_->[1],    #commits
                    l => $l          #label
                    }

            } splice( @{$response}, $min, ( $max + 1 ) )
            }
        : { map {
                my $w = wday( $_->[0] );
                my $l = label( $_->[1] );
                push( @{ $self->{ca} }, $GitInsight::Util::CA_COLOURS{$l} );
                $last = $l if ( !$last );

                #   $commits_count{ $_->[1] } = 1;
                $self->{stats}->{$w}->{$l}++
                    if $self->statistics == 1;    #filling stats hashref
                $self->{transition_hash}->{$w}->{$last}
                    ->{$l}++;                     #filling stats hashref
                $self->{transition}->{$w}
                    ->slice("$last,$l")++;        #filling transition matrix
                $last = $l;
                $_->[0] => {
                    c => $_->[1],                 #commits
                    d => $w,                      #day in the week
                    l => $l                       #label
                    }

            } splice( @{$response}, $min, ( $max + 1 ) )
        }
    );

    return $self->contribs;
}

sub process {
    my $self = shift;
    croak "process() called while you have not specified an username"
        if !$self->username;
    $self->contrib_calendar( $self->username )
        if !$self->contribs and $self->username;
    $self->_transition_matrix;
    $self->_markov;
    $self->_gen_stats if ( $self->statistics );
    $self->{png} = $self->draw_ca( @{ $self->{ca} } )
        if ( $self->ca_output == 1 );
    $self->{steps} = \%GitInsight::Util::LABEL_STEPS;
    $self->_accuracy if $self->accuracy and $self->accuracy == 1;
    return $self;
}

sub _gen_stats {
    my $self = shift;
    my $sum  = 0;
    if ( $self->no_day_stats ) {
        $sum += $_ for values %{ $self->{stats} };
        foreach my $k ( keys %{ $self->{stats} } ) {
            info "Calculating probability for label $k  $sum /  "
                . $self->{stats}->{$k}
                if $self->verbose;
            my $prob = prob( $sum, $self->{stats}->{$k} );
            info "Is: $prob" if $self->verbose;
            $self->{stats}->{$k} = sprintf "%.5f", $prob;
        }
    }
    else {
        foreach my $k ( keys %{ $self->{stats} } ) {
            $sum = 0;
            $sum += $_ for values %{ $self->{stats}->{$k} };
            map {
                info "Calculating probability for $k -> label $_  $sum /  "
                    . $self->{stats}->{$k}->{$_}
                    if $self->verbose;
                my $prob = prob( $sum, $self->{stats}->{$k}->{$_} );
                info "Is: $prob" if $self->verbose;
                $self->{stats}->{$k}->{$_} = sprintf "%.5f", $prob;
            } ( keys %{ $self->{stats}->{$k} } );
        }
    }
}

sub _markov {
    my $self = shift;
    info "Markov chain phase" if $self->verbose;
    my $dayn = 1;
    info "Calculating predictions for "
        . ( scalar( @{ $self->{last_week} } ) ) . " days"
        if $self->verbose;

    foreach my $day ( @{ $self->{last_week} } ) {    #cycling the last week
        my $wd = wday( $day->[0] );                  #computing the weekday
        my $ld = $day->[1];                          #getting the label
        my $M  = markov_list(
            gen_m_mat($ld),
            $self->no_day_stats
            ? $self->{transition}
            : $self->{transition}->{$wd},
            $dayn
        );    #Computing the markov for the state

        my $label = 0;
        $M->[$label] > $M->[$_] or $label = $_ for 1 .. scalar(@$M) - 1;
        push( @{ $self->{ca} }, $GitInsight::Util::CA_COLOURS{$label} )
            ;    #adding the predictions to ca

        my ( $mday, $mon, $year )
            = reverse( split( /-/, $day->[0] ) );    #splitting date

        push(
            @{ $self->{result} },
            [   $wd, $label,
                $day->[0] = strftime(
                    '%Y-%m-%d',
                    localtime(
                        timelocal( 0, 0, 0, $mday, $mon - 1, $year )
                            + 7 * 86_400
                    )
                    ) #adding 7 days to the date, and adding the result to $self->{result}
                ,
                $M
            ]
        );

        if ( $self->verbose ) {
            info "$wd: "
                . $label . " has "
                . ( sprintf "%.2f", $M->[$label] * 100 )
                . "% of probability to happen";
            info "\t"
                . $_
                . " ---- "
                . ( sprintf "%.2f", $M->[$_] * 100 ) . "%"
                for 0 .. scalar(@$M) - 1;
        }

        ############# TREEMAP GENERATION #############
        $self->{'treemap'}->{'name'} = "day";
        my $hwd = { name => $day->[0], children => [] };
        push(
            @{ $hwd->{children} },
            { name => $_, size => $M->[$_] * 10000 }
        ) for 0 .. scalar(@$M) - 1;
        push( @{ $self->{'treemap'}->{"children"} }, $hwd );
        ################################################

        $dayn++ if $self->no_day_stats;
    }

    return $self->{result};

}

sub _transition_matrix {

#transition matrix, sum all the transitions occourred in each day,  and do prob(sumtransitionrow ,current transation occurrance )
    my $self = shift;
    info "Going to build transation matrix probabilities" if $self->verbose;
    if ( $self->no_day_stats ) {
        my $sum = $self->{transition}->sumover();
        map {
            foreach my $c ( 0 .. LABEL_DIM ) {
                $self->{transition}->slice("$_,$c")
                    .= prob( # slice of the single element of the matrix , calculating bayesian inference
                    $sum->at($c),    #contains the transition sum of the row
                    $self->{transition}->at( $_, $c )
                    );    # all the transation occurred, current transation
            }
        } ( 0 .. LABEL_DIM );
    }
    else {
        foreach my $k ( keys %{ $self->{transition} } ) {
            my $sum = $self->{transition}->{$k}->sumover();
            map {
                foreach my $c ( 0 .. LABEL_DIM ) {
                    $self->{transition}->{$k}->slice("$_,$c")
                        .= prob( # slice of the single element of the matrix , calculating bayesian inference
                        $sum->at($c)
                        , #contains the transition sum of the row over the day
                        $self->{transition}->{$k}->at( $_, $c )
                        )
                        ; # all the transation occurred in those days, current transation
                }
            } ( 0 .. LABEL_DIM );
        }
    }
}

1;
__END__

=encoding utf-8

=head1 NAME

GitInsight - Predict your github contributions using Bayesian inference and Markov chain

=head1 SYNOPSIS

    gitinsight --username [githubusername] (--nodaystats) (--accuracy) #using the shipped bin

    #or using the module

    my $Insight= GitInsight->new(no_day_stats => 0, username => "markov", accuracy=> 1);
    my $Result= $Insight->process;
    my $accuracy = $Insight->{accuracy};
    $Result = $Insight->{result};
    # $Result contains the next week predictions and is an arrayref of arrayrefs    [  [ 'Sat', 1, '2014-07-1', [ 0 ,  '0.151515151515152', '0.0606060606060606', '0.0404040404040404',  0  ]  ],   ..   [            'DayofWeek',      'winner_label',  'day' ,  [             probability_label_0,  probability_label_1,              probability_label_2,          probability_label_3,              probability_label_4            ]          ]]


=head1 DESCRIPTION

GitInsight is module that allow you to predict your github contributions in the "calendar contribution" style of github (the table of contribution that you see on your profile page).

=head1 HOW DOES IT WORK?

GitInsight generates a transation probrability matrix from your github contrib_calendar to compute the possibles states for the following days. Given that GitHub split the states thru 5 states (or here also called label), the probability can be inferenced by using Bayesian methods to update the beliefs of the possible state transition, while markov chain is used to predict the states. The output of the submitted data is then plotted using Cellular Automata.

=head2 THEORY

We trace the transitions states in a matrix and increasing the count as far as we observe a transition (L<https://en.wikipedia.org/wiki/Transition_matrix>), then we inference the probabilities using Bayesan method L<https://en.wikipedia.org/wiki/Bayesian_inference> L<https://en.wikipedia.org/wiki/Examples_of_Markov_chains>.

=head1 INSTALLATION

GitInsight requires the installation of gsl (GNU scientific library), gd(http://libgd.org/), PDL and PDL::Stats  (to be installed after the gsl library set).

on Debian:

        apt-get install gsl-bin libgs10-devt
        apt-get install pdl libpdl-stats-perl libgd2-xpm-dev

It's reccomended to use cpanm to install all the required deps, install it thru your package manager or just do:

    cpan App::cpanminus

After the installation of gsl, clone the repository and install all the dependencies with cpanm:

    cpanm --installdeps .

Then install it as usual:

    perl Build.PL
    ./Build
    ./Build test #ensure that the module works correctly
    ./Build install

=head1 OPTIONS

=head2 username

required, it's the GitHub username used to calculate the prediction

=head2 ca_output

you can enable/disable the cellular autmata output using this option (1/0)

=head2 no_day_stats

setting this option to 1, will slightly change the prediction: it will be calculated a unique transition matrix instead one for each day

=head2 left_cutoff

used to cut the days from the start (e.g. if you want to delete the first 20 days from the prediction, just set this to 20)

=head2 cutoff_offset

used to select a range where the prediction happens (e.g. if you want to calculate the prediction of a portion of your year of contribution)

=head2 file_output

here you can choose the file output name for ca.

=head2 accuracy

Enable/disable accuracy calculation (1/0)

=head2 verbose

Enable/disable verbosity (1/0)

=head1 METHODS

=head2 contrib_calendar($username)

Fetches the github contrib_calendar of the specified user

=head2 process

Calculate the predictions and generate the CA

=head2 start_day

Returns the first day of the contrib_calendar

=head2 last_day

Returns the last day of the contrib calendar (prediction included)

=head2 prediction_start_day

Returns the first day of the prediction (7 days of predictions)

=head1 AUTHOR

mudler E<lt>mudler@dark-lab.netE<gt>

=head1 COPYRIGHT

Copyright 2014- mudler

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<GitInsight::Util>, L<PDL>, L<PDL::Stats>

=cut
